sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -114,9 +129,93 @@
         }
         this.tenantId = tenantId;
         this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.certificatePath = certificatePath;
+        this.certificatePassword = certificatePassword;
         this.options = options;
     }
 
+    private ConfidentialClientApplication getConfidentialClientApplication() {
+        if (confidentialClientApplication != null) {
+            return confidentialClientApplication;
+        } else if (clientId == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException(
+                "A non-null value for client ID must be provided for user authentication."));
+        } else {
+            String authorityUrl = options.getAuthorityHost().replaceAll("/+$", "") + "/" + tenantId;
+            IClientCredential credential;
+            if (clientSecret != null) {
+                credential = ClientCredentialFactory.createFromSecret(clientSecret);
+            } else if (certificatePath != null) {
+                try {
+                    if (certificatePassword == null) {
+                        byte[] pemCertificateBytes = Files.readAllBytes(Paths.get(certificatePath));
+                        credential = ClientCredentialFactory.createFromCertificate(
+                            CertificateUtil.privateKeyFromPem(pemCertificateBytes),
+                            CertificateUtil.publicKeyFromPem(pemCertificateBytes));
+                    } else {
+                        credential = ClientCredentialFactory.createFromCertificate(
+                            new FileInputStream(certificatePath), certificatePassword);
+                    }
+                } catch (CertificateException
+                    | UnrecoverableKeyException
+                    | NoSuchAlgorithmException
+                    | KeyStoreException
+                    | NoSuchProviderException
```

Comment:
Thoughts on merging all these to use their super class `GeneralSecurityException`.

Reply:
works like a charm